### ISSUE TITLE
Port SpatialPandas Hilbert-Rtree implementation

### ISSUE THREAD

**Issue by tastatham** [tastatham/gsoc_dask_geopandas_2021#5]
Initial work has begun to port SpatialPanda Hilbert-Rtree implementation into Dask-GeoPandas.
This can be located in [dask_geopandas_wip](https://github.com/tastatham/gsoc_dask_geopandas_2021/tree/master/functions/dask_geopandas_wip).
SpatialPandas uses metadata to store two attributes that are used to calculate the hilbert-rtree;
* Partition bounds (_partition_bounds)
* Spatial index (_sindex)
In [dask_geopandas_wip](https://github.com/tastatham/gsoc_dask_geopandas_2021/tree/master/functions/dask_geopandas_wip)., I have created one new attribute - partition bounds but I have reused the spatial_partitions meta.
All changes can be seen on lines containing # MODIFIED or by running a diff.

Initial work is shown in [comparing_hilbert_rtree.ipynb](https://github.com/tastatham/gsoc_dask_geopandas_2021/blob/master/notebooks/dask_geopandas/comparing_hilbert_rtree.ipynb). 

Steps:
- [ ] Port SpatialPandas into Dask-GeoPandas wip repo
- [ ] Create tests
- [ ] Benchmark
- [ ] Refactor Hilbert-Rtree

**Comment by jorisvandenbossche** [comment]
The Hilbert implementation from spatialpandas consists of two parts, right? One file that calculates the distances along the hilbert curve, and another file that implements an rtree spatial index on top of this. So we maybe could split this task in multiple steps.

Personally, I would start with only the "hilbert_distance", and add a function in dask-geopandas to compute the hilbert distance for all geometries in the GeoSeries. The repartition function of spatialpandas (they call it `pack_partitions`) only uses the distance AFAIK. (https://github.com/holoviz/spatialpandas/blob/9252a7aba5f8bc7a435fffa2c31018af8d92942c/spatialpandas/dask.py#L172-L200)
I am bit skeptical that we actually need the second step (the tree based on this), since we now have the fast strtree from pygeos, but that's still something we can certainly check and compare in a second step.

**Comment by martinfleis** [comment]
Agree with @jorisvandenbossche here. 

Let's not complicate our lives with new models of spatial indexing. Reusing the pygeos-based machinery we have in geopandas is the path of least resistance and with multithreading issue being fixed, there's no significant reason to avoid it.

**Comment by tastatham** [comment]
Yes, I agree to keep things simple for now and calculate just the hilbert distance.
I hadn't looked into exactly how SpatialPandas re-partitions (it's where I got up to in the comparing_hilbert_rtree.ipynb) the data but knew it used the `pack_partitions` function.

Looking at that function, it looks like it does not use the rtree for repartitioning. 
Looking into the internals of the `pack_partitions` function, the [ `_distances_from_bounds` function inside the rtree.py](https://github.com/holoviz/spatialpandas/blob/9252a7aba5f8bc7a435fffa2c31018af8d92942c/spatialpandas/spatialindex/rtree.py#L50) is used - which is kind of confusing.
It's probably there because `_distances_from_bounds` function is used by the HilbertRtree class and the hilbertcurve was refactored from another package.

We can discuss this in the meeting today.
